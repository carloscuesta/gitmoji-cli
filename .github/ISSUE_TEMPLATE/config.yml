blank_issues_enabled: false
contact_links:
  - name: 😍 Contribution Guide
    url: https://github.com/carloscuesta/gitmoji-cli/blob/master/.github/CONTRIBUTING.md
    about: Please read through before making any contribution.