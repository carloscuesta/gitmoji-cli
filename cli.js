#!/usr/bin/env node
'use strict';

const meow = require('meow');
const axios = require('axios');
const updateNotifier = require('update-notifier');
const GitmojiCli = require('./src/gitmoji.js');
const pkg = require('./package.json');

updateNotifier({pkg}).notify();

const cli = meow(`
	Usage
	  $ gitmoji
	Options
		--init, -i	Initialize gitmoji as a commit hook
		--commit, -c Interactively commit using the prompts
		--list, -l  List all the available gitmojis
		--search, -s	Search gitmojis
		--version, -v	Print gitmoji-cli installed version
		--update, -u	Sync emoji list with the repo
	Examples
		$ gitmoji -l
		$ gitmoji bug linter -s
`, {
	alias: {
		i: 'init',
		c: 'commit',
		l: 'list',
		s: 'search',
		h: 'help',
		v: 'version',
		u: 'update'
	}
});

const gitmojiApiClient = axios.create({
	baseURL: 'https://raw.githubusercontent.com/carloscuesta/gitmoji/master',
	timeout: 5000,
	headers: {},
	params: {}
});

const gitmojiCli = new GitmojiCli(gitmojiApiClient);

const commands = {
	list: () => gitmojiCli.list(),
	search: () => cli.input.map(element => gitmojiCli.search(element)),
	init: () => gitmojiCli.init(),
	hook: () => gitmojiCli.ask('hook'),
	version: () => console.log(gitmojiCli.version(pkg.version)),
	commit: () => gitmojiCli.ask('client'),
	update: () => gitmojiCli.updateCache(),
	undefined: () => {
		if (process.argv[2] === '--hook') {
			gitmojiCli.ask('hook');
		} else {
			gitmojiCli.ask('client');
		}
	}
};

const arg = Object.keys(cli.flags)[1];
if (commands[arg]) {
	commands[arg]();
} else {
	cli.showHelp();
}
