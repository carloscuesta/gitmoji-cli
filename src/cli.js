#!/usr/bin/env node
const meow = require('meow')
const updateNotifier = require('update-notifier')

const pkg = require('../package.json')
const GitmojiCli = require('./gitmoji.js')
const utils = require('./utils.js')
const commands = require('./commands').default

updateNotifier({ pkg }).notify()

const cli = meow(
  `
  Usage
    $ gitmoji
  Options
    --commit, -c    Interactively commit using the prompts
    --config, -g    Setup gitmoji-cli preferences.
    --init, -i      Initialize gitmoji as a commit hook
    --list, -l      List all the available gitmojis
    --remove, -r    Remove a previously initialized commit hook
    --search, -s    Search gitmojis
    --update, -u    Sync emoji list with the repo
    --version, -v   Print gitmoji-cli installed version
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
`,
  {
    flags: {
      commit: { type: 'boolean', alias: 'c' },
      config: { type: 'boolean', alias: 'g' },
      help: { type: 'boolean', alias: 'h' },
      init: { type: 'boolean', alias: 'i' },
      list: { type: 'boolean', alias: 'l' },
      remove: { type: 'boolean', alias: 'r' },
      search: { type: 'boolean', alias: 's' },
      update: { type: 'boolean', alias: 'u' },
      version: { type: 'boolean', alias: 'v' }
    }
  }
)

const gitmojiCli = new GitmojiCli(utils.gitmojiApiClient)
const options = {
  commit: () => gitmojiCli.ask('client'),
  config: () => gitmojiCli.config(),
  hook: () => gitmojiCli.ask('hook'),
  init: () => gitmojiCli.init(),
  list: () => gitmojiCli.list(),
  remove: () => gitmojiCli.remove(),
  search: () => cli.input.map((input) => commands.search(input)),
  update: () => gitmojiCli.updateCache()
}

utils.findGitmojiCommand(cli, options)
