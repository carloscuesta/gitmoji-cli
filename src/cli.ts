#!/usr/bin/env node
import meow from 'meow'
import updateNotifier from 'update-notifier'
import { readFileSync } from 'fs'

import FLAGS from '@constants/flags'
import findGitmojiCommand from '@utils/findGitmojiCommand'
import type { CommitOptions } from '@commands/commit'
import type { SearchOptions } from '@commands/search'

const packageJson = readFileSync(
  new URL('../package.json', import.meta.url)
).toString()

updateNotifier({ pkg: JSON.parse(packageJson) }).notify({
  isGlobal: true
})

const cli = meow(
  `
  Usage
    $ gitmoji [option] [command]
  Options
    --${FLAGS.COMMIT}, -c    Interactively commit using the prompts
    --${FLAGS.CONFIG}, -g    Setup gitmoji-cli preferences.
    --${FLAGS.INIT}, -i      Initialize gitmoji as a commit hook
    --${FLAGS.LIST}, -l      List all the available gitmojis
    --${FLAGS.REMOVE}, -r    Remove a previously initialized commit hook
    --${FLAGS.SEARCH}, -s    Search gitmojis
    --${FLAGS.UPDATE}, -u    Sync emoji list with the repo
    --${FLAGS.VERSION}, -v   Print gitmoji-cli installed version
  Commands
    commit          Interactively commit using the prompts
    config          Setup gitmoji-cli preferences.
    init            Initialize gitmoji as a commit hook
    list            List all the available gitmojis
    remove          Remove a previously initialized commit hook
    search          Search gitmojis
    update          Sync emoji list with the repo
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
`,
  {
    importMeta: { url: import.meta.url } as ImportMeta,
    flags: {
      [FLAGS.COMMIT]: { type: 'boolean', alias: 'c' },
      [FLAGS.CONFIG]: { type: 'boolean', alias: 'g' },
      [FLAGS.HELP]: { type: 'boolean', alias: 'h' },
      [FLAGS.INIT]: { type: 'boolean', alias: 'i' },
      [FLAGS.LIST]: { type: 'boolean', alias: 'l' },
      [FLAGS.REMOVE]: { type: 'boolean', alias: 'r' },
      [FLAGS.SEARCH]: { type: 'boolean', alias: 's' },
      [FLAGS.UPDATE]: { type: 'boolean', alias: 'u' },
      [FLAGS.VERSION]: { type: 'boolean', alias: 'v' }
    }
  }
)

export const options = {
  [FLAGS.COMMIT]: async (options: CommitOptions) =>
    await (await import('@commands/commit')).default(options),
  [FLAGS.CONFIG]: async () =>
    await (await import('@commands/config')).default(),
  [FLAGS.HOOK]: async (options: CommitOptions) =>
    await (await import('@commands/commit')).default(options),
  [FLAGS.INIT]: async () =>
    await (await import('@commands/hook')).default.create(),
  [FLAGS.LIST]: async () => await (await import('@commands/list')).default(),
  [FLAGS.REMOVE]: async () =>
    await (await import('@commands/hook')).default.remove(),
  [FLAGS.SEARCH]: async (options: SearchOptions) =>
    await (await import('@commands/search')).default(options),
  [FLAGS.UPDATE]: async () => await (await import('@commands/update')).default()
}

findGitmojiCommand(cli, options)
