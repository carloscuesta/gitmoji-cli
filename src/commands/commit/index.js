// @flow
import inquirer from 'inquirer'

import getEmojis from '../../utils/getEmojis'
import prompts from './prompts'
import withHook from './withHook'
import withClient from './withClient'

const commit = (mode: 'client' | 'hook') => {
  return getEmojis()
    .then((gitmojis) => prompts(gitmojis))
    .then((questions) => {
      inquirer.prompt(questions).then((answers) => {
        if (mode === 'hook') return withHook(answers)

        return withClient(answers)
      })
    })
}

export default commit
