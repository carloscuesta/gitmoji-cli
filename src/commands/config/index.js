// @flow
import inquirer from 'inquirer'

import configurationPrompts, { CONFIGURATION_PROMPT_NAMES } from './prompts'
import configurationVault from '../../utils/configurationVault'

const config = () => {
  inquirer.prompt(configurationPrompts()).then((answers) => {
    configurationVault.setAutoAdd(answers[CONFIGURATION_PROMPT_NAMES.AUTO_ADD])
    configurationVault.setEmojiFormat(
      answers[CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT]
    )
    configurationVault.setSignedCommit(
      answers[CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT]
    )
    configurationVault.setScopePrompt(
      answers[CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT]
    )
  })
}

export default config
