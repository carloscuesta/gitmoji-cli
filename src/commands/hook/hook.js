// @flow
const HOOK: Object = {
  PERMISSIONS: 0o775,
  PATH: '/hooks/prepare-commit-msg',
  CONTENTS: `#!/bin/sh
# gitmoji as a commit hook
commit_message_empty=true
while read -r line
do
  case "$line" in
  "# ------------------------ >8 ------------------------"*)
    break
    ;;
  ""|"#"*)
    # skip
    ;;
  *)
    commit_message_empty=false
    break
    ;;
  esac
done < "$1"
if test -t 1 && $commit_message_empty; then
  # it has been invoked from a tty and no commit message is already set
  exec < /dev/tty
  gitmoji --hook "$1"
fi`
}

export default HOOK
