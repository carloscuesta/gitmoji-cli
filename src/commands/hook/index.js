import create from './create'
import remove from './remove'

export default {
  create,
  remove
}
