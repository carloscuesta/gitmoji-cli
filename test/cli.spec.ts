import updateNotifier from 'update-notifier'
import meow from 'meow'

import commit from '@commands/commit/index.js'
import config from '@commands/config/index.js'
import hook from '@commands/hook/index.js'
import list from '@commands/list/index.js'
import search from '@commands/search/index.js'
import update from '@commands/update/index.js'

import pkg from '../package.json'
import { options } from '../src/cli.js'

jest.mock('@commands/commit/index.js')
jest.mock('@commands/config/index.js')
jest.mock('@commands/hook/index.js')
jest.mock('@commands/list/index.js')
jest.mock('@commands/search/index.js')
jest.mock('@commands/update/index.js')

describe('cli', () => {
  it('should call updateNotifier', () => {
    expect(updateNotifier).toHaveBeenCalledWith({ pkg })
    expect(updateNotifier({ pkg }).notify).toHaveBeenCalledWith({
      isGlobal: true
    })
  })

  it('should match meow with cli information', () => {
    const mockMeow = meow as jest.Mock
    mockMeow.mock.calls[0][1].importMeta = 'import.meta.url'
    expect(mockMeow.mock.calls).toMatchSnapshot()
  })

  it('should call commit command on commit', async () => {
    await options.commit({ mode: 'client' })
    expect(commit).toHaveBeenCalledWith({ mode: 'client' })
  })

  it('should call config command on config', async () => {
    await options.config()
    expect(config).toHaveBeenCalledWith()
  })

  it('should call commit command on hook', async () => {
    await options.hook({ mode: 'hook' })
    expect(commit).toHaveBeenLastCalledWith({ mode: 'hook' })
  })

  it('should call createHook command on init', async () => {
    await options.init()
    expect(hook.create).toHaveBeenCalledWith()
  })

  it('should call list command on list', async () => {
    await options.list()
    expect(list).toHaveBeenCalledWith()
  })

  it('should call removeHook command on remove', async () => {
    await options.remove()
    expect(hook.remove).toHaveBeenCalledWith()
  })

  it('should call search command on search', async () => {
    await options.search({ query: ['testSearchQuery'] })
    expect(search).toHaveBeenCalledWith({ query: ['testSearchQuery'] })
  })

  it('should call update command on update', async () => {
    await options.update()
    expect(update).toHaveBeenCalledWith()
  })
})
