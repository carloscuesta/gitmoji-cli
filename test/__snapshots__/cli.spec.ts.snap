// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should match meow with cli information 1`] = `
[
  [
    "
  Usage
    $ gitmoji [option] [command]
  Options
    --commit, -c    Interactively commit using the prompts
    --config, -g    Setup gitmoji-cli preferences.
    --init, -i      Initialize gitmoji as a commit hook
    --list, -l      List all the available gitmojis
    --remove, -r    Remove a previously initialized commit hook
    --search, -s    Search gitmojis
    --update, -u    Sync emoji list with the repo
    --version, -v   Print gitmoji-cli installed version
  Commands
    commit          Interactively commit using the prompts
    config          Setup gitmoji-cli preferences.
    init            Initialize gitmoji as a commit hook
    list            List all the available gitmojis
    remove          Remove a previously initialized commit hook
    search          Search gitmojis
    update          Sync emoji list with the repo
  Examples
    $ gitmoji -l
    $ gitmoji bug linter -s
",
    {
      "flags": {
        "commit": {
          "shortFlag": "c",
          "type": "boolean",
        },
        "config": {
          "shortFlag": "g",
          "type": "boolean",
        },
        "help": {
          "shortFlag": "h",
          "type": "boolean",
        },
        "init": {
          "shortFlag": "i",
          "type": "boolean",
        },
        "list": {
          "shortFlag": "l",
          "type": "boolean",
        },
        "remove": {
          "shortFlag": "r",
          "type": "boolean",
        },
        "search": {
          "shortFlag": "s",
          "type": "boolean",
        },
        "update": {
          "shortFlag": "u",
          "type": "boolean",
        },
        "version": {
          "shortFlag": "v",
          "type": "boolean",
        },
      },
      "importMeta": "import.meta.url",
    },
  ],
]
`;
